widgetry/src/backend_glow_native.rs

```diff
@@ -89,3 +91,202 @@ impl WindowAdapter {
         self.0.swap_buffers().unwrap();
     }
 }
+
+/// Uploads a sprite sheet of textures to the GPU so they can be used by FancyColor::Texture and
+/// friends to paint shapes.
+///
+/// `path` - image file which is a grid of images.
+/// `sprite_width` - the width and height of an individual cell in the image grid
+///
+/// The image file can have any number of sprites, but they must all be the same size.
+///
+/// Once uploaded, textures are addressed by their id, starting from 1, from left to right, top to
+/// bottom, like so:
+///
+///     ┌─┬─┬─┐
+///     │1│2│3│
+///     ├─┼─┼─┤
+///     │4│5│6│
+///     ├─┼─┼─┤
+///     │7│8│9│
+///     └─┴─┴─┘
+///
+/// Texture(0) is reserved for a pure white (no-op) texture.
+///
+/// Implementation is based on the the description of ArrayTextures from:
+/// https://www.khronos.org/opengl/wiki/Array_Texture.
+fn load_textures(
+    gl: &glow::Context,
+    filename: &str,
+    sprite_width: u32,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let path = abstutil::path(filename);
+    let image_bytes = abstutil::slurp_file(&path)?;
+    let dynamic_img = image::load_from_memory(&image_bytes)?;
+    let img = if let image::DynamicImage::ImageRgba8(img) = dynamic_img {
+        img
+    } else {
+        todo!("support other image formats");
+    };
+
+    let format = glow::RGBA;
+    let target = glow::TEXTURE_2D_ARRAY;
+    let mipmap_level = 1;
+    let internal_format = glow::RGBA;
+
+    let texture_id = unsafe { gl.create_texture()? };
+    unsafe {
+        gl.bind_texture(target, Some(texture_id));
+    }
+
+    let (img_width, img_height) = img.dimensions();
+    let sprite_height = sprite_width;
+    let sprites_per_row = img_width / sprite_width;
+    let sprites_per_column = img_height / sprite_width;
+    let sprite_count = sprites_per_row * sprites_per_column;
+
+    assert_eq!(
+        sprites_per_row * sprite_width,
+        img_width,
+        "sprites must align exactly"
+    );
+    assert_eq!(
+        sprites_per_column * sprite_height,
+        img_height,
+        "sprites must align exactly"
+    );
+
+    info!(
+        "img_size: {}x{}px ({} px), sprite_size: {}x{}px, sprites: {}x{} ({} sprites)",
+        img_width,
+        img_height,
+        img.pixels().len(),
+        sprite_width,
+        sprite_height,
+        sprites_per_row,
+        sprites_per_column,
+        sprite_count
+    );
+
+    // In order to avoid branching in our shader logic, all shapes are rendered with a texture.
+    // Even "non-textured" styles like FancyColor::RGBA, use a "default" no-op (pure white) texture,
+    // which we generate here.
```

Comment:
Please note this @dabreegster. This is how I avoided any branching logic in the shader logic.

It might be controversial. I don't know what's common practice, all I have to go off of are some random forum posts where people said this was "a fine way to do it". 🤷 

So our shader can do this in all cases (even for "untextured" stuff like FancyColor::RGBA:

```
// in the no-op texture case, texture(textures, texture_id) -> #ffffffff, so the multiplication is a no-op
fs_color = color * texture(textures, texture_id);
```

The alternative would be something like this in the shader:
```
if texture_id == 0 {
    fs_color = blah
} else {
    fs_color = color * texture(textures, texture_id);
}
```